Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -683,19 +683,30 @@ void Device::draw_primitives(PrimitiveType primitive_type, FloatMatrix4x4 const&
 
                     FloatVector4 vertex_to_light;
 
+                    if ((vertex.eye_coordinates.w() != 0.0f) && (light.position.w() == 0.0f)) {
+                        vertex_to_light = light.position;
+                    } else if ((vertex.eye_coordinates.w() == 0.0f) && (light.position.w() != 0.0f)) {
+                        vertex_to_light = -vertex.eye_coordinates;
+                    } else {
+                        // FIXME: The OpenGL 1.5 spec says (in typical SGI fashion) _nothing_ about if P1 and P2 have a w value of 1, which would
```

Comment:
do you mean a w value of 0? since the else case of both the above ifs is either both are 0 or both are something else

Reply:
This should say _both_. I'll fix the typo.